lib/ansible/modules/crypto/openssl_certificate.py

```diff
@@ -589,35 +627,156 @@ def _validate_csr():
+        self._module = module
+
+        try:
+            self.privatekey = crypto_utils.load_privatekey(
+                self.privatekey_path, self.privatekey_passphrase, backend=self.backend
+            )
+        except crypto_utils.OpenSSLBadPassphraseError as exc:
+            module.fail_json(msg=to_native(exc))
+
+    def generate(self, module):
+        if not os.path.exists(self.privatekey_path):
+            raise CertificateError(
+                'The private key %s does not exist' % self.privatekey_path
+            )
+        if not os.path.exists(self.csr_path):
+            raise CertificateError(
+                'The certificate signing request file %s does not exist' % self.csr_path
+            )
+        if self.digest is None:
+            raise CertificateError(
+                'The digest %s is not supported with the cryptography backend' % module.params['selfsigned_digest']
+            )
+
+        if not self.check(module, perms_required=False) or self.force:
+            try:
+                cert_builder = x509.CertificateBuilder()
+                cert_builder = cert_builder.subject_name(self.csr.subject)
+                cert_builder = cert_builder.issuer_name(self.csr.subject)
+                cert_builder = cert_builder.serial_number(self.serial_number)
+                cert_builder = cert_builder.not_valid_before(self.notBefore)
+                cert_builder = cert_builder.not_valid_after(self.notAfter)
+                cert_builder = cert_builder.public_key(self.privatekey.public_key())
+                for extension in self.csr.extensions:
+                    cert_builder = cert_builder.add_extension(extension.value, critical=extension.critical)
+            except ValueError as e:
+                raise CertificateError(str(e))
+
+            certificate = cert_builder.sign(
+                private_key=self.privatekey, algorithm=self.digest,
+                backend=default_backend()
+            )
+
+            self.cert = certificate
+
+            try:
+                with open(self.path, 'wb') as cert_file:
+                    cert_file.write(certificate.public_bytes(Encoding.PEM))
+            except Exception as exc:
+                raise CertificateError(exc)
+
+            self.changed = True
+        else:
+            self.cert = crypto_utils.load_certificate(self.path, backend=self.backend)
+
+        file_args = module.load_file_common_arguments(module.params)
+        if module.set_fs_attributes_if_different(file_args, False):
+            self.changed = True
+
+    def dump(self, check_mode=False):
+
+        result = {
+            'changed': self.changed,
+            'filename': self.path,
+            'privatekey': self.privatekey_path,
+            'csr': self.csr_path
+        }
+
+        if check_mode:
+            result.update({
+                'notBefore': self.notBefore,
+                'notAfter': self.notAfter,
```

Comment:
notBefore and notAfter should be converted to strings (ASN.1), as here they are `datetime` objects.